font/src/directwrite/mod.rs

```diff
@@ -107,73 +232,42 @@ impl crate::Rasterize for DirectWriteRasterizer {
+        self.keys.insert(desc.clone(), key);
+        self.fonts.insert(key, LoadedFont { font, font_face });
 
-        Ok(FontKey { token: (self.fonts.len() - 1) as u16 })
+        Ok(key)
     }
+}
 
-    fn get_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
-        let font = self.fonts.get(glyph.font_key.token as usize).ok_or(Error::FontNotLoaded)?;
-
-        let offset = GlyphOffset { advanceOffset: 0.0, ascenderOffset: 0.0 };
-
-        let glyph_index = *font
-            .get_glyph_indices(&[glyph.c as u32])
-            .first()
-            .ok_or_else(|| Error::MissingGlyph(glyph.c))?;
-        if glyph_index == 0 {
-            // The DirectWrite documentation states that we should get 0 returned if the glyph
-            // does not exist in the font
-            return Err(Error::MissingGlyph(glyph.c));
-        }
-
-        let glyph_run = dwrote::DWRITE_GLYPH_RUN {
-            fontFace: unsafe { font.as_ptr() },
-            fontEmSize: glyph.size.as_f32_pts(),
-            glyphCount: 1,
-            glyphIndices: &(glyph_index),
-            glyphAdvances: &(0.0),
-            glyphOffsets: &(offset),
-            isSideways: 0,
-            bidiLevel: 0,
-        };
+impl crate::Rasterize for DirectWriteRasterizer {
+    type Err = Error;
 
-        let rendering_mode = font.get_recommended_rendering_mode_default_params(
-            glyph.size.as_f32_pts(),
-            self.device_pixel_ratio * (96.0 / 72.0),
-            dwrote::DWRITE_MEASURING_MODE_NATURAL,
-        );
+    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
+        Ok(DirectWriteRasterizer {
+            fonts: HashMap::new(),
+            keys: HashMap::new(),
+            device_pixel_ratio,
+            font_collection: FontCollection::system(),
+            font_fallback: dwrote::FontFallback::get_system_fallback(),
+        })
+    }
 
-        let glyph_analysis = GlyphRunAnalysis::create(
-            &glyph_run,
-            self.device_pixel_ratio * (96.0 / 72.0),
-            None,
-            rendering_mode,
-            dwrote::DWRITE_MEASURING_MODE_NATURAL,
-            0.0,
-            0.0,
-        )
-        .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
+        let loaded_font = self.fonts.get(&key).ok_or(Error::FontNotLoaded)?;
+        self.font_metrics(&loaded_font.font_face, size)
+    }
 
-        let bounds = glyph_analysis
-            .get_alpha_texture_bounds(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1)
-            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
-        let buf = glyph_analysis
-            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
-            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+    fn load_font(&mut self, desc: &FontDesc, _size: Size) -> Result<FontKey, Error> {
+        self.get_matching_face(desc)
```

Comment:
It seems like this `get_matching_face` function is also only ever called from here? Then we can probably inline it.